ortools/sat/csharp/SearchHelpers.cs

```diff
@@ -31,50 +31,66 @@ public class CpSolverSolutionCallback : SolutionCallback
      */
     public long Value(LinearExpr e)
     {
-        List<Term> terms = new List<Term>();
-        terms.Add(new Term(e, 1));
         long constant = 0;
+        long coefficient = 1;
+        LinearExpr expr = e;
+        Queue<Term> terms = CpSolver.RentQueue();
```

Comment:
Just add a Queue field in the callback. Only initialize it on demand. By construction, only one callback can be executed at a given time.